File: hht-alpha-integrator/environment/reference/pendulum_ref.py
#!/usr/bin/env python3
"""
Minimal-coordinate ODE reference for a planar double pendulum
(uniform rigid rods, pin joints, gravity).

Uses classical 4th-order Runge-Kutta with a fine time step (1e-4) for
high-accuracy reference trajectories.  No external dependencies beyond
the Python standard library.

Usage:
  python3 pendulum_ref.py theta1 theta2 omega1 omega2 t_end output.csv

Physical parameters match the C++ model defaults:
  L1 = L2 = 1, m1 = m2 = 1, g = 9.81
"""

import math
import sys

L1, L2 = 1.0, 1.0
M1, M2 = 1.0, 1.0
G = 9.81


def rhs(y):
    """Right-hand side of the double pendulum ODE in minimal coordinates."""
    theta1, theta2, omega1, omega2 = y
    c = math.cos(theta1 - theta2)
    s = math.sin(theta1 - theta2)

    a11 = (M1 / 3.0 + M2) * L1 * L1
    a12 = 0.5 * M2 * L1 * L2 * c
    a22 = M2 * L2 * L2 / 3.0

    b1 = (-0.5 * M2 * L1 * L2 * s * omega2 * omega2
          - G * math.cos(theta1) * (M1 * L1 / 2.0 + M2 * L1))
    b2 = (0.5 * M2 * L1 * L2 * s * omega1 * omega1
          - M2 * G * (L2 / 2.0) * math.cos(theta2))

    det = a11 * a22 - a12 * a12
    return [omega1, omega2,
            (a22 * b1 - a12 * b2) / det,
            (a11 * b2 - a12 * b1) / det]


def integrate(y0, t_end, h=1e-4):
    """RK4 integration with output at 0.001-second intervals."""
    n = int(round(t_end / h))
    out_dt = min(0.001, t_end)
    out_every = max(1, int(round(out_dt / h)))

    y = list(y0)
    times = [0.0]
    states = [list(y)]

    for i in range(1, n + 1):
        k1 = rhs(y)
        y2 = [y[j] + 0.5 * h * k1[j] for j in range(4)]
        k2 = rhs(y2)
        y3 = [y[j] + 0.5 * h * k2[j] for j in range(4)]
        k3 = rhs(y3)
        y4 = [y[j] + h * k3[j] for j in range(4)]
        k4 = rhs(y4)
        y = [y[j] + h / 6.0 * (k1[j] + 2 * k2[j] + 2 * k3[j] + k4[j])
             for j in range(4)]

        if i % out_every == 0:
            times.append(i * h)
            states.append(list(y))

    if abs(times[-1] - t_end) > 1e-12:
        times.append(n * h)
        states.append(list(y))

    return times, states


def main():
    if len(sys.argv) != 7:
        sys.exit("Usage: pendulum_ref.py theta1 theta2 omega1 omega2 t_end output.csv")

    theta1 = float(sys.argv[1])
    theta2 = float(sys.argv[2])
    omega1 = float(sys.argv[3])
    omega2 = float(sys.argv[4])
    t_end = float(sys.argv[5])
    out_file = sys.argv[6]

    times, states = integrate([theta1, theta2, omega1, omega2], t_end)

    with open(out_file, "w") as f:
        f.write("time,theta1,theta2,omega1,omega2\n")
        for t, s in zip(times, states):
            f.write(f"{t:.15e},{s[0]:.15e},{s[1]:.15e},"
                    f"{s[2]:.15e},{s[3]:.15e}\n")


if __name__ == "__main__":
    main()

File: hht-alpha-integrator/environment/reference/pendulum_ref.m
% pendulum_ref.m  —  minimal-coordinate ODE reference for a planar double
% pendulum (uniform rigid rods, pin joints, gravity).
%
% Usage:
%   octave-cli --no-gui pendulum_ref.m theta1 theta2 omega1 omega2 t_end output.csv
%
% Physical parameters are hardcoded to match the C++ model defaults:
%   L1 = L2 = 1, m1 = m2 = 1, g = 9.81

args = argv();
if length(args) ~= 6
  error('Need 6 args: theta1 theta2 omega1 omega2 t_end output_file');
end

theta1_0  = str2double(args{1});
theta2_0  = str2double(args{2});
omega1_0  = str2double(args{3});
omega2_0  = str2double(args{4});
t_end_val = str2double(args{5});
out_file  = args{6};

L1 = 1.0;  L2 = 1.0;
m1 = 1.0;  m2 = 1.0;
g_val = 9.81;

y0 = [theta1_0; theta2_0; omega1_0; omega2_0];

opts = odeset('RelTol', 1e-12, 'AbsTol', 1e-14);
[t_out, y_out] = ode45( ...
    @(t, y) pend_rhs(t, y, L1, L2, m1, m2, g_val), ...
    [0, t_end_val], y0, opts);

fid = fopen(out_file, 'w');
fprintf(fid, 'time,theta1,theta2,omega1,omega2\n');
for i = 1:length(t_out)
  fprintf(fid, '%.15e,%.15e,%.15e,%.15e,%.15e\n', ...
          t_out(i), y_out(i,1), y_out(i,2), y_out(i,3), y_out(i,4));
end
fclose(fid);

% -----------------------------------------------------------------------
function ydot = pend_rhs(~, y, L1, L2, m1, m2, g)
  theta1 = y(1);  theta2 = y(2);
  omega1 = y(3);  omega2 = y(4);

  c = cos(theta1 - theta2);
  s = sin(theta1 - theta2);

  A11 = (m1 / 3.0 + m2) * L1^2;
  A12 = 0.5 * m2 * L1 * L2 * c;
  A22 = m2 * L2^2 / 3.0;

  b1 = -0.5 * m2 * L1 * L2 * s * omega2^2 ...
       - g * cos(theta1) * (m1 * L1 / 2.0 + m2 * L1);
  b2 =  0.5 * m2 * L1 * L2 * s * omega1^2 ...
       - m2 * g * (L2 / 2.0) * cos(theta2);

  d = A11 * A22 - A12^2;
  alpha1 = (A22 * b1 - A12 * b2) / d;
  alpha2 = (A11 * b2 - A12 * b1) / d;

  ydot = [omega1; omega2; alpha1; alpha2];
endfunction
